### PR TITLE
Use server-side copy for restore from S3 to S3.

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Improvement

### Changelog entry:
If a backup and restoring data are both in S3 then server-side copy should be used from now on.

Server side means that if your backup and a table you're restoring both are located in S3 then during the restore process data will be copied only inside S3 server without using network much. Today data is always first downloaded from S3 to clickhouse-server (by portions), then uploaded back to S3.

### CODE REVIEW

File: src/Backups/BackupImpl.cpp [hunk 1]
```diff
@@ -677,45 +607,140 @@ BackupEntryPtr BackupImpl::readFile(const SizeAndChecksum & size_and_checksum) c
...
+
+    std::unique_ptr<SeekableReadBuffer> read_buffer;
+    std::unique_ptr<SeekableReadBuffer> base_read_buffer;
+
+    if (info.size > info.base_size)
     {
         /// Data goes completely from this backup, the base backup isn't used.
-        return std::make_unique<BackupEntryFromBackupImpl>(
-            std::static_pointer_cast<const BackupImpl>(shared_from_this()), info.archive_suffix, info.data_file_name, info.size, info.checksum);
+        if (use_archives)
```
[the remark this answers is not in the document]
Comment (vitlibar) [reply src/Backups/BackupImpl.cpp:624]: Yes, I've edited it.

File: src/Backups/BackupIO_File.cpp [hunk 2]
```diff
@@ -30,6 +32,19 @@ std::unique_ptr<SeekableReadBuffer> BackupReaderFile::readFile(const String & fi
...
+bool BackupReaderFile::supportNativeCopy(DataSourceDescription destination_data_source_description, WriteMode mode) const
+{
+    return (destination_data_source_description == getDataSourceDescription()) && (mode == WriteMode::Rewrite);
+}
+
+void BackupReaderFile::copyFileToDiskNative(const String & file_name, size_t, DiskPtr destination_disk, const String & destination_path, WriteMode)
+{
+    std::string abs_source_path = path / file_name;
+    std::string abs_destination_path = fullPath(destination_disk, destination_path);
+    fs::copy(abs_source_path, abs_destination_path, fs::copy_options::overwrite_existing);
```
[the remark this answers is not in the document]
Comment (vitlibar) [reply src/Backups/BackupIO_File.cpp]: I searched through our code base, we mostly use `fs::copy()`. The difference doesn't seem to be big anyway. `fs::copy()` follows symlinks, I think that's useful but it doesn't seem really important anyway.

**Review by antonio2368 — APPROVED** [review]

**Comment by alifirat** [comment]
Hi, can you clarify what do you mean by server-side ? 
What will be different regarding today ?

**Comment by vitlibar** [comment]
> Hi, can you clarify what do you mean by server-side ? What will be different regarding today ?

Server side means that if your backup and a table you're restoring both are located in S3 then during the restore process data will be copied only inside S3 server without using network much. Today data is always first downloaded from S3 to clickhouse-server (by portions), then uploaded back to S3.

**Comment by alifirat** [comment]
Ok thanks for the clarification @vitlibar just want to be sure :)

**Pull request merged**